### ISSUE TITLE
Honour - Possible DOS on Voter::createGuage for non-standard velocimeter pools

### ISSUE THREAD

**Issue by sherlock-admin3** [sherlock-audit/2024-06-velocimeter-judging#422]
Honour

Medium

# Possible DOS on Voter::createGuage for non-standard velocimeter pools

## Summary

Voter::createGauge incorrectly assumes that all pairs (including non-velocimeter pairs) have an externalBribe or adheres to the IPair interface.

## Vulnerability Detail

Voter::createGauge allows the governor/emergencyCouncil to create gauges for non-velocimeter pairs, however, non-velocimeter pools cannot be expected to have an externalBribe or adhere to the IPair interface , as a result the function call on [line 354](https://github.com/sherlock-audit/2024-06-velocimeter/blob/main/v4-contracts/contracts/Voter.sol#L354-L354) would cause a revert in such cases. [ln 372-375](https://github.com/sherlock-audit/2024-06-velocimeter/blob/main/v4-contracts/contracts/Voter.sol#L372-L375) shows that the externalBribe is only added to compliant pools anyways.

## Impact

DOS on Voter::createGuage for non-standard velocimeter pools

## Code Snippet

https://github.com/sherlock-audit/2024-06-velocimeter/blob/main/v4-contracts/contracts/Voter.sol#L354-L354

## Tool used

Manual Review

## Recommendation

```diff

  function createGauge(address _pool, uint256 _gaugeType) external returns (address) {
      require(_gaugeType < factories.length, "gaugetype big");
      require(gauges[_pool] == address(0x0), "exists");
      address[] memory allowedRewards = new address[](3);
      address[] memory internalRewards = new address[](2);
      address tokenA;
      address tokenB;
      address _factory = factories[_gaugeType];
      address _gaugeFactory = gaugeFactories[_gaugeType];
      require(_factory != address(0));
      require(_gaugeFactory != address(0));
      bool isPair = IPairFactory(_factory).isPair(_pool);
+     address _external_bribe ;

      if (isPair) {
          tokenA = IPair(_pool).token0();
          tokenB = IPair(_pool).token1();
          allowedRewards[0] = tokenA;
          allowedRewards[1] = tokenB;
          internalRewards[0] = tokenA;
          internalRewards[1] = tokenB;
          // if one of the tokens is not base (FLOW) then add base(FLOW) to allowed rewards
          if (base != tokenA && base != tokenB) {
            allowedRewards[2] = base;
          }
+         _external_bribe = IPair(_pool).externalBribe();
+         if(_external_bribe == address(0)) {
+           _external_bribe = IBribeFactory(bribefactory).createExternalBribe(allowedRewards); 
+         }
      }

      if (msg.sender != governor && msg.sender != emergencyCouncil) { // gov can create for any pool, even non-Velocimeter pairs
          require(isPair, "!_pool");
          require(IGaugePlugin(gaugePlugin).checkGaugeCreationAllowance(msg.sender, tokenA, tokenB), "!whitelistedForGaugeCreation");
      }

-     address _external_bribe = IPair(_pool).externalBribe();
-     
-     if(_external_bribe == address(0)) {
-       _external_bribe = IBribeFactory(bribefactory).createExternalBribe(allowedRewards); 
-     }
      address _gauge = IGaugeFactory(_gaugeFactory).createGauge(_pool, _external_bribe, _ve, isPair, allowedRewards);

      IERC20(base).approve(_gauge, type(uint).max);
      external_bribes[_gauge] = _external_bribe;
      gauges[_pool] = _gauge;
      poolForGauge[_gauge] = _pool;
      isGauge[_gauge] = true;
      isAlive[_gauge] = true;
      _updateFor(_gauge);
      if(claimable[_gauge] > 0) {
          claimable[_gauge] = 0;
      }
      pools.push(_pool);
      if (isPair) {
          IPair(_pool).setHasGauge(true);
          IPair(_pool).setExternalBribe(_external_bribe);
      }
      emit GaugeCreated(_gauge, msg.sender, _external_bribe, _pool);
      return _gauge;
  }
```

Duplicate of #579

**Comment by Honour-d-dev** [comment]
Escalate

This issue and #579 are valid
Non-velocimeter pools cannot be expected to have to have an external bribe. And the function implementation [here](https://github.com/sherlock-audit/2024-06-velocimeter/blob/main/v4-contracts/contracts/Voter.sol#L372-L375) shows that the external bribe is indeed not required if the pool is non-velocimeter.  
Hence this is actually a valid issue as gauge creation can be Dos'd by an unnecessary functionality

**Comment by sherlock-admin3** [comment]
> Escalate
> 
> This issue and #579 are valid
> Non-velocimeter pools cannot be expected to have to have an external bribe. And the function implementation [here](https://github.com/sherlock-audit/2024-06-velocimeter/blob/main/v4-contracts/contracts/Voter.sol#L372-L375) shows that the external bribe is indeed not required if the pool is non-velocimeter.  
> Hence this is actually a valid issue as gauge creation can be Dos'd by an unnecessary functionality

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by nevillehuang** [comment]
@cvetanovv For this issue please see discussion [here](https://discord.com/channels/812037309376495636/1272953615413153833) for my thoughts. Will loop in sponsor to review as well @dawiddrzala

**Comment by cvetanovv** [comment]
I think this issue is valid because `createGuage()` cannot create non-standard velocimeter pools. 

In the code, as a comment, it is written that the function should be able to create non-standard pools:

https://github.com/sherlock-audit/2024-06-velocimeter/blob/63818925987a5115a80eff4bd12578146a844cfd/v4-contracts/contracts/Voter.sol#L349
> gov can create for any pool, **even** non-Velocimeter pairs

Planning to accept the escalation and make this issue a valid Medium severity.

**Comment by nevillehuang** [comment]
@cvetanovv This is a questionable decision, because who is to say the non-velocimeter pair cannot implement the same functionalities, when this contracts are not even in scope and such decisions to confirm the code logic will be made by the admin?

**Comment by Honour-d-dev** [comment]
@nevillehuang it shows [here](https://github.com/sherlock-audit/2024-06-velocimeter/blob/63818925987a5115a80eff4bd12578146a844cfd/v4-contracts/contracts/Voter.sol#L372C9-L375C10) that the external bribe is not even set regardless of whether the non-velocimeter pool implements the functionality or not. Which implies it's not a required functionality and shouldn't dos pool that don't have it
```solidity
if (isPair) {
            IPair(_pool).setHasGauge(true);
            IPair(_pool).setExternalBribe(_external_bribe);
        }
```

**Comment by nevillehuang** [comment]
@Honour-d-dev Imo this is not a good argument, because ultimately, you are still making speculations and assumptions on what is the oos non-velocimeter pair contract functionalities.

**Comment by Honour-d-dev** [comment]
i believe it's the reverse. `createGauge` is assuming functionalities of an oos non-velocimeter pair and this report is saying it shouldn't, no?

**Comment by nevillehuang** [comment]
@Honour-d-dev I can have a OOS wrapper contract and/or completely new OOS non-velocimeter Pair contract that the protocol can craft having the external bribes functions required doing nothing and I will avoid the revert, so there are arguably no issues. Again this would involve possibility of OOS logic, either way I see it as invalid. I will refrain from any further comments and respect hoj decision unless necessary

**Comment by Honour-d-dev** [comment]
Yeah, but this is just assuming OOS functionalities/logic, it's better to not check for anything we don't intend to use and eliminate all possible failure points

**Comment by dawiddrzala** [comment]
all of the pools require to have the external bribe as well this is technical requiment, even not pair ones need to have external bribe so users can vote on it bribe etc, only admin can create custom gauges (not pair once etc for example proxy gauge ), so admin is doing the verification , so this is not valid issue ( your proposed fix break the proxy gauge for example )

**Comment by Honour-d-dev** [comment]
@dawiddrzala appreciate the response ,it is true that external bribes are required for voting and also that the fix(on this specific report) does actually break voting as it creates the possibility of the external bribe for the gauge to be address(0).

However, this issue's group should still be considered valid for these reasons:
- The fix proposed by #579 does not break anything (i can provide a POC for this if required)
- Although the external bribe is a required functionality (especially for voting on a pair) ,It's shown in multiple instances throughout the code that it is not required for the pair itself to implement this feature if it's non-velocimeter  
[`_setExternalBribe`](https://github.com/sherlock-audit/2024-06-velocimeter/blob/63818925987a5115a80eff4bd12578146a844cfd/v4-contracts/contracts/Voter.sol#L148C5-L153C6)
```solidity
function _setExternalBribe(address _gauge, address _external) private {
        external_bribes[_gauge] = _external;
        address pool = poolForGauge[_gauge];
->      try IPair(pool).setExternalBribe(_external){} catch {}
        emit ExternalBribeSet(msg.sender, _gauge, _external);
    }
```
[`createGauge`#L372-L375](https://github.com/sherlock-audit/2024-06-velocimeter/blob/63818925987a5115a80eff4bd12578146a844cfd/v4-contracts/contracts/Voter.sol#L372C9-L375C10)
```solidity
        if (isPair) {
            IPair(_pool).setHasGauge(true);
            IPair(_pool).setExternalBribe(_external_bribe);
        }
```
In summary i believe this vulnerability exists and [#L354](https://github.com/sherlock-audit/2024-06-velocimeter/blob/63818925987a5115a80eff4bd12578146a844cfd/v4-contracts/contracts/Voter.sol#L354) can actually Dos createGauge in such cases (of course without assuming any OOS functionalities like a wrapper) and the fix by #579 should be used instead.

**Comment by cvetanovv** [comment]
Considering this sponsor's [comment](https://github.com/sherlock-audit/2024-06-velocimeter-judging/issues/422#issuecomment-2294602019) and the fact that the admin is Trusted to do his check, this issue is invalid.

Planning to reject the escalation and leave the issue as is.

**Comment by pronobis4** [comment]
I also think that everything is designed in such a way that not "any" pool can be added. If you add a wrapper, it is not "any" pool, but it is tailored to the project.

That is why I suggested a simple fix, which is often used in the project for such cases.

**Comment by Honour-d-dev** [comment]
@pronobis4 i think you meant "designed in such a way that "any" pool can be added" instead of
>I also think that everything is designed in such a way that _not_ "any" pool can be added

For example admin can decide to add any uniswap/curve pool and it just works. The extensive use of try/catch also suggests this. From the current code it won't be possible and we'll have to assume OOS functionalities (like a wrapper) will be used.

To be honest i think this is just a simple fix that prevents a lot of overhead. The decision is left for Hoj

**Comment by cvetanovv** [comment]
As the sponsor wrote, Admin is Trusted and will do his checks. My decision to reject the escalation remains.

**Comment by WangSecurity** [comment]
Result:
Invalid
Duplicate of #579

**Comment by sherlock-admin4** [comment]
Escalations have been resolved successfully!

Escalation status:
- [Honour-d-dev](https://github.com/sherlock-audit/2024-06-velocimeter-judging/issues/422/#issuecomment-2286840601): rejected
